numba/targets/arraymath.py

```diff
@@ -517,6 +517,24 @@ def flat_any(a):
 
     return flat_any
 
+@overload(np.average)
+def np_average(arr, weights=None):
+    if weights is None or isinstance(weights, types.NoneType):
+        def np_average_impl(arr, weights=None):
+            return np.mean(arr)
+
+    else:
+        def np_average_impl(arr, weights=None):
+            weighted_arr = np.multiply(arr, weights)
```

Comment:
What if `arr` and `weights` are dimensionally incompatible? Also, are there cases where broadcasting may happen?

Reply:
Yes, broadcasting can happen if the `weights` array is 1D, but the `arr` array is multidimensional. I'm having trouble implementing this because I'm not sure how to set up the `weights` array so that it can be broadcast. For example, if we have an `arr` with shape (3,2,1) and `weights` with shape (2,) then I think we would want to reshape `weights` from (2,) to (1,2,1). This is done in the NumPy implementation: https://github.com/numpy/numpy/blob/7ee52003/numpy/lib/function_base.py#L415-L417.

When I try using `(arr.ndim-1)*(1,) + weights.shape) ` (which is part of `np.broadcast_to(weights, (arr.ndim-1)*(1,) + weights.shape)`, line 416 of the NumPy implementation) to create a tuple (i.e. (1,2,1)) so that I can change the shape of the weights array from (2,) to (1,2,1) (e.g. `weights.reshape((1,2,1))`). I get an error stating that this function is not implemented:
```
    TypingError: Failed in nopython mode pipeline (step: nopython frontend)
Invalid use of Function(<built-in function mul>) with argument(s) of type(s): (int64, tuple(int64 x 1))
Known signatures:
 * (int64, int64) -> int64
 * (int64, uint64) -> int64
 * (uint64, int64) -> int64
 * (uint64, uint64) -> uint64
 * (float32, float32) -> float32
 * (float64, float64) -> float64
 * (complex64, complex64) -> complex64
 * (complex128, complex128) -> complex128
 * parameterized
In definition 0:
    All templates rejected with literals.
In definition 1:
    All templates rejected without literals.
In definition 2:
    All templates rejected with literals.
In definition 3:
    All templates rejected without literals.
In definition 4:
    All templates rejected with literals.
In definition 5:
    All templates rejected without literals.
In definition 6:
    All templates rejected with literals.
In definition 7:
    All templates rejected without literals.
```
Is there another way to create a tuple (i.e (1,2,1) in this case) in the `np_average_impl` function?

Reply:
I've taken a bit of a closer look at this, the [documentation](https://docs.scipy.org/doc/numpy-1.15.0/reference/generated/numpy.average.html) says:
> weights : array_like, optional
An array of weights associated with the values in a. Each value in a contributes to the average according to its associated weight. The weights array can either be 1-D (in which case its length must be the size of a along the given axis) or of the same shape as a. If weights=None, then all data in a are assumed to have a weight equal to one.

It seems to outline that `weights` can be 1-D conditional upon a `given axis` being present, trying this out:
```python
In [24]: np.average(np.arange(24).reshape(2,3,4), weights=np.arange(4))                                                         
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-24-46ab532e92a4> in <module>
----> 1 np.average(np.arange(24).reshape(2,3,4), weights=np.arange(4))

<path>/lib/python3.7/site-packages/numpy/lib/function_base.py in average(a, axis, weights, returned)
    368             if axis is None:
    369                 raise TypeError(
--> 370                     "Axis must be specified when shapes of a and weights "
    371                     "differ.")
    372             if wgt.ndim != 1:

TypeError: Axis must be specified when shapes of a and weights differ.

In [25]: np.average(np.arange(24).reshape(2,3,4), weights=np.arange(4), axis=2)                                                 
Out[25]: 
array([[ 2.33333333,  6.33333333, 10.33333333],
       [14.33333333, 18.33333333, 22.33333333]])
```
demonstrates this to be the case.

I think in the case of the current PR, perhaps it's best to not support the `axis` kwarg (this is a common omission in Numba) and to raise an exception if the shape of the `weights` array does not match the shape of the `a` array, this providing consistency with NumPy?